src/machines/java/com/enderio/machines/common/io/fluid/MachineFluidTank.java

```diff
@@ -0,0 +1,244 @@
+        else if (action.simulate()) {
+            if (fluid.isEmpty()) return Math.min(capacity, source.getAmount());
+            else return Math.min(capacity - fluid.getAmount(), source.getAmount());
+        } else {
+            if (fluid.isEmpty()) {
+                fluid = new FluidStack(source, Math.min(capacity, source.getAmount()));
+                onContentsChanged();//case source is empty is checked in beginning.
+                return fluid.getAmount();
+            } else {
+                int availableSpace = capacity - fluid.getAmount();
+                int sourceAmount = source.getAmount();
+                int transferAmount = Math.min(availableSpace,sourceAmount);
+                fluid.grow(transferAmount);
+                //according to the parent implementation the source shall not be drained
+                if (transferAmount>0)
+                    onContentsChanged();
+                return transferAmount;
+            }
+        }
+    }
+
+    //Same as parent behavior with the addition of this class capabilities, see method 'fill' above for details.
+    public int fill(int desiredAmount, FluidAction action, boolean force){
+        if(fluid.isEmpty()){
+            EnderIO.LOGGER.error("No fluid in tank, can't contain an amount of unspecified fluid other than 0");
+        }
+        return fill(new FluidStack(fluid.getFluid(),desiredAmount), action,force);
+    }
+    //Same as parent behavior with the addition of this class capabilities, see method 'fill' above for details.
+    @Override
+    public int fill(FluidStack source, FluidAction action) {
+        return fill(source,action, false);
+    }
+
+    /** drains fluid out of the tank.
+     * @param maxDrain - The maximal amount that you wish to draw in mb. In case of using a bucket it would be FluidType.BUCKET_VOLUME (1000)mb.
+     * @param action - execute if the tank amount should be changed, simulate if not.
+     * @param force - Force your way around the allow output flag. Usually used to drain a machine "input" tank from within the machine.
+     * @return - The amount that was drained in case of action is execute and how much could be drawn in case of
+     */
+    public int drain(int maxDrain, FluidAction action, boolean force){
+        if(maxDrain <= 0) return 0;
+        if(!allowOutput && !force) return 0; // Continue draining only if either allow drain or force
+        int transferAmount = Math.min(maxDrain, fluid.getAmount());
+        if(action == FluidAction.EXECUTE && transferAmount>0){
+            fluid.shrink(transferAmount);
+            onContentsChanged();
+        }
+        return transferAmount;
+    }
+
+    //Same as parent behavior with the addition of this class capabilities, see method 'drain' above for details.
+    @Override
+    public @NotNull FluidStack drain(FluidStack resource, FluidAction action) {
+        if (resource.isEmpty() || !resource.isFluidEqual(fluid)) return FluidStack.EMPTY;
```

Comment:
put return on new line and don't use else as it's the only other path as it return early

Reply:
Done